### ISSUE TITLE
As a system administrator I want to have easy to parse log messages. If I configure JSON as log format the whole log message should be formatted as json objects.

### ISSUE THREAD

**Issue by stefan-kalb** [keycloak/keycloak#31860]
As a system administrator I want to have easy to parse log messages. If I configure JSON as log format the whole log message should be formatted as json objects.

Currently the contents of the json object "**message**" are no json object but an escaped text message.
Example JSON log message:

```
{
  "timestamp": "2024-08-02T14:26:35.2639447+02:00",
  "sequence": 9838,
  "loggerClassName": "org.jboss.logging.Logger",
  "loggerName": "org.keycloak.events",
  "level": "WARN",
  "message": "type=\"LOGIN_ERROR\", realmId=\"4b939ad4-eaa4-489a-a818-81fb5ce3c8de\", realmName=\"test\", clientId=\"44ff6470-44c1-436a-aede-66745bdb4b3e\", userId=\"null\", ipAddress=\"172.17.99.100\", error=\"user_not_found\", auth_method=\"openid-connect\", redirect_uri=\"https://test.testserver.lan/oauth/login\", code_id=\"2faeafc2-784e-4e71-9f88-eaad6cb303cd\", username=\"test\"",
  "threadName": "executor-thread-1",
  "threadId": 28,
  "mdc": {},
  "ndc": "",
  "hostName": "lab-dock-10.lab.lan",
  "processName": "QuarkusEntryPoint",
  "processId": 2649028
}

```

The content of "message" should be also a json object, like:

```
{
  "timestamp": "2024-08-02T14:26:35.2639447+02:00",
  "sequence": 9838,
  "loggerClassName": "org.jboss.logging.Logger",
  "loggerName": "org.keycloak.events",
  "level": "WARN",
  "message": {
    "type": "LOGIN_ERROR",
    "realmId": "4b939ad4-eaa4-489a-a818-81fb5ce3c8de",
    "realmName": "test",
    "clientId": "44ff6470-44c1-436a-aede-66745bdb4b3e",
    "userId": "null",
    "ipAddress": "172.17.99.100",
    "error": "user_not_found",
    "auth_method": "openid-connect",
    "redirect_uri": "https://test.testserver.lan/oauth/login",
    "code_id": "2faeafc2-784e-4e71-9f88-eaad6cb303cd",
    "username": "test"
  },
  "threadName": "executor-thread-1",
  "threadId": 28,
  "mdc": {},
  "ndc": "",
  "hostName": "lab-dock-10.lab.lan",
  "processName": "QuarkusEntryPoint",
  "processId": 2649028
}
```

_Originally posted by @stefan-kalb in https://github.com/keycloak/keycloak/discussions/31859_

**Comment by stefan-kalb** [comment]
I hope I did as intended in opening a feature request issue alongside a discussion.

**Comment by shawkins** [comment]
> I hope I did as intended in opening a feature request issue alongside a discussion.

Hopefully a separate discussion won't be needed because this is a fairly narrow enhancement request.

What you are asking for seems to be some additional configuration so that the logging of the event is done in json, instead of the custom comma separate list of name value pairs done here https://github.com/keycloak/keycloak/blob/7c69c857a1216cf2af07c4d2859fc961817f530a/services/src/main/java/org/keycloak/events/log/JBossLoggingEventListenerProvider.java#L83

Alternatively you could use a custom logging provider for this.

Then you want the Quarkus / Jboss logging JsonFormatter to inline messages that are already json, rather than escaping them as a string https://github.com/quarkusio/quarkus/blob/b2699a501517e7aebb928d097e74cd55b2003c8f/extensions/logging-json/runtime/src/main/java/io/quarkus/logging/json/runtime/JsonFormatter.java#L11

This second part seems expensive to determine, and may not strictly be necessary as you could just parse the message as json once the first part is done.

**Comment by stefan-kalb** [comment]
Thanks. I already do the key-value parsing of the included text messages.

However imo it would be a useful feature to format the textual key-value pairs as json key-value pairs for json logging. In that case a syslog parser would be able to parse the json string in a single step.

**Comment by shawkins** [comment]
Deleted the discussion to keep things centralized. 

> Thanks. I already do the key-value parsing of the included text messages.

Based upon the initial assessment, and this feedback, marking this as low priority.

If this gets closed due to lack of activity / up-votes, then consider converting to a discussion for further refinement until something actionable is created.

~priority-low

**Comment by oculos** [comment]
This really would be important for us. Performance-wise it would make things much smoother.

**Comment by mabartos** [comment]
@shawkins There's already some SPI for providing additional log entries on the Quarkus side, which might be helpful even here for the formatting? Would be nice to investigate -> https://quarkus.io/blog/quarkus-3-36-released/#dynamic-json-field-additions-for-logging

**Comment by shawkins** [comment]
Thanks for the pointer @mabartos - the docs are at https://quarkus.io/guides/logging#json-provider

That could be used with our next release, but would be unsupported. 

However since that still just has access to the log record, you would need to probably still tweak what the JBoss Logging event listener is producing.
